[8.0] product_extended_segmentation: enhance segmentation update message logs

osvalr opened the pull request:
### Task [4989](https://www.vauxoo.com/web#id=4989&view_type=form&model=project.task&menu_id=136&action=138)
### This solves #799
## Changes included:
- [X] Write messages directly to bom's related products instead of product's template.
- [X] Include an assertion in one of its tests in order to take in count the inclusion of this message.
- [x] Unify in one single method both logging calls from `product_extended_variants` and `product_extended_segmentation` when product gets written
  - `product_extended_variants` wasn't taken in count but wizard used in `product.product` to updated 
    price
  - Write log when:
    - Cost updated, Segments got updated
    - Cost updated, Segments didn't
    - Cost NOT updated, Segments got updated
    - Cost NOT updated, Segments didn't
- [x] Set percent variation up two decimal precision
- [x] Include in unit tests:
  - [x] messages written
  - [x] price value variation

---
- [x] rebased

---
### Current behavior:

> ![image](https://cloud.githubusercontent.com/assets/2961943/14795396/1ad96c9c-0af2-11e6-9458-7320ae3f57e6.png)
> [![](http://img.youtube.com/vi/ebmNbi-dMnI/0.jpg)](https://youtu.be/ebmNbi-dMnI)
> [![](http://img.youtube.com/vi/gRmzNVQCJgg/0.jpg)](https://youtu.be/gRmzNVQCJgg)
### Behavior these changes introduces:

> ![image](https://cloud.githubusercontent.com/assets/2961943/14769178/c99114f0-0a1c-11e6-83ca-bb26bd6b448e.png)
> [![](http://img.youtube.com/vi/-DJFM31bQpk/0.jpg)](https://youtu.be/-DJFM31bQpk)

hbto commented on product_extended_segmentation/model/template.py:
```diff
@@ -216,15 +215,19 @@ def _factor(factor, product_efficiency, product_rounding):
                 (current_price and diff < 0 and
                     computed_th > bottom_price_threshold):
-            tmpl_obj.message_post(cr, uid, [product_tmpl_id.id],
+            product_id = prod_obj.search(cr, uid, [('bom_ids', '=', bom.id)])
```
By doing this search you could be excluding all those product.product records that are not linked to a BoM record
use criteria ('product_tmpl_id', '=', product_tmpl_id.id)

Best Regards

osvalr replied:
I've taken this in count as we discussed earlier

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -216,15 +215,20 @@ def _factor(factor, product_efficiency, product_rounding):
                     computed_th > bottom_price_threshold):
-            tmpl_obj.message_post(cr, uid, [product_tmpl_id.id],
+            product_ids = prod_obj.search(
+                cr, uid, [('product_tmpl_id', '=', product_tmpl_id.id)])
```
I do not understand call me please.

This is being executed in the context of a product, why a search?

osvalr replied:
Yes it's product context, I just changed to use mapped instead.

osvalr commented:
Hi @nhomar! I'll be uploading a video for this change, but please review it

nhomar commented:
Can you rebase it and left travis green please.... (before the video)

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -247,10 +237,82 @@ def _factor(factor, product_efficiency, product_rounding):
+                '{0}'.format(str(sgmnts_dict))
+        else:
+            body = ''
+            subject = "Segments NOT updated. "
```
put body and subject outside and before the if, even if the else looks like is right the proper way is initialize the variables with its default value and then execute the change over them (in favor of readability):

``` python
body = 'Default Body'
subject = 'Subject default'
if something:
   ....
   body = new_body
   subject = new_subject
```

osvalr replied:
sure

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -247,10 +237,82 @@ def _factor(factor, product_efficiency, product_rounding):
+            if not log_only:
+                prod_ids.write(new_price)
+            subject += "Cost updated correctly. "
+            body += ""
```
I can not understand why are you summing chars instead build them? can you enlgiht me please

osvalr replied:
They're just literals, that's why I add them up

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -201,35 +197,29 @@ def _factor(factor, product_efficiency, product_rounding):
+        }
+        if threshold_reached:
+            return self.ensure_change_price_log_messages(cr, uid, vals,
+                                                         context)
```
mucho más limpio +1

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -247,10 +237,82 @@ def _factor(factor, product_efficiency, product_rounding):
+            (self.standard_price and diff < 0 and
+             computed_th > bottom_th)
+
+    def _compute_threshold(self, std_price, new_price):
```
puedes quitar el self y usat un @staticmethod

te queadría así:

``` python
@staticmethod
def _compute_threshold(std_price, new_price):
   .... do what you do without erach a cr,,,
   return whatever
```

Commit:
[IMP] product_extended_segmentation: write log in bom's related product

 - define methods for threshold
 - include unit test
 - include wizard calling method for product and evaluate threshold
before continue
 - Write on product logs instead of template
 - use product records from context instead values pass by args
 - set percent variation up to 2 decimals precision
 - update readme with new features

osvalr commented:
@nhomar, I took in count your suggestions also I've restarted the runbot's dummy pull

nhomar commented on product_extended_segmentation/model/template.py:
```diff
@@ -247,10 +237,82 @@ def _factor(factor, product_efficiency, product_rounding):
+        new_price = vals.get('new_price', False)
+        prod_ids = vals.get('product_ids', False)
+        if prod_ids._name == 'product.template':
+            prod_ids = prod_ids.product_variant_ids
```
this evaluation is not comming in the context?

Why are you asking for the _name?

I was waiting for the self._context.get('active_model') comparison, here it will send always to all prod_ids as I understand.

osvalr replied:
you were right I simplify the code using `self.env.context` ids

nhomar commented:
LEt me know when update the video with the last behavior please.

osvalr commented:
@nhomar I've updated the video, watch it and let me know your comments about it

osvalr commented on product_extended_segmentation/README.rst:
```diff
@@ -4,16 +4,41 @@
+* Set a bottom threshold for those price updates that make the product price
+decrease, the default value is 0 (No decrease allowed).
+* All update price wizards for products check if new cost meets the
+bottom threshold limit.
```
@nhomar if you believe some semantic here is not right, let me know

osvalr commented on product_extended_segmentation/README.rst:
```diff
@@ -4,16 +4,41 @@
 ======================
 
-* No issues have been reported yet.
+* Top threshold limit is not currently supported.
```
I think this is needed to leave it clear
